File: Hash Table/zero_sum_subarray.go
/*
	You're given a list of integers nums. Write a function that returns a boolean representing
	whether there exists a zero-sum subarray of nums

	Sample Input :  = [-5, -5, 2, 3, -2]
	Output : True
	The subarray [-5, 2, 3] has a sum of 0

    Approach:
    Time and Space complexity : O(n) time | O(n) space - where n is the length of nums
    This implementation uses an unordered map to keep track of the prefix sum of the input array nums. 
    We initialize the map with a key-value pair of 0 and -1, since a prefix sum of 0 indicates that the 
    subarray from index 0 to -1 (i.e., an empty subarray) has a sum of 0. We then iterate through the 
    input array nums, adding each element to the running sum sum and checking if the current sum is 
    already in the map. If it is, then we've found a subarray whose sum is 0, so we add the starting and 
    ending indices of the subarray to the result vector and break out of the loop.
    If we reach the end of the loop without finding a zero sum subarray, then we return an empty vector.
    Note that this implementation assumes that there is only one zero sum subarray in the input array. 
    If there could be multiple zero sum subarrays, then we would need to modify the implementation to 
    return all of them.
*/
package main

import "fmt"

func ZeroSumSubarray(nums []int) bool {
	sums := map[int]bool{0: true}
    currentSum := 0
    for _, num := range nums {
        currentSum += num
        if _, sumIsInSet := sums[currentSum]; sumIsInSet {
            return true
        }
        sums[currentSum] = true
    }
    return false
}

func main() {
	nums := []int{-5, -5, 2, 3, -2}
	fmt.Println(ZeroSumSubarray(nums))
	nums = []int{1, 2, 3, 4, 5}
	fmt.Println(ZeroSumSubarray(nums))
}

File: Hash Table/ZeroSumSubarray.java
/*
Q -> Hash Table: You're given a list of integers nums.
Write a function that returns a boolean representing whether there exists a zero-sum subarray of nums in Java
*/

//Here's a Java code that implements a solution to find if there is a zero-sum subarray in the given list of
// integers using a hash table.

import java.util.*;

public class ZeroSumSubarray {

    public static boolean hasZeroSumSubarray(int[] nums) {
        Set<Integer> set = new HashSet<>();
        int sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            if (sum == 0 || set.contains(sum)) {
                return true;
            }
            set.add(sum);
        }
        return false;
    }

    public static void main(String[] args) {
        int[] nums = {4, -3, 2, 1, 8};
        boolean hasZeroSumSubarray = hasZeroSumSubarray(nums);
        System.out.println(hasZeroSumSubarray);
    }
}

/*
Explanation:-

The above code uses a HashSet to keep track of the prefix sum of the elements of the input array. At each index, it checks if the sum up to that index has already been seen before in the HashSet.
If it has been seen, it means that there exists a subarray with a zero sum. If the sum equals zero, then the subarray starts from the beginning of the array.
The time complexity of this algorithm is O(n), where n is the size of the input array, since we traverse the input array only once.
The space complexity is also O(n), since the HashSet can contain up to n elements in the worst case.
 */
